### PR TITLE
Throw if options is changed

### LINKED ISSUE #1018: redefining value is possible
This parses just fine in BOUT.inp
```
[neutrals]
type=diffusion
type=parallel
```
Is this intended to have the same value defined twice?
Should there be a warning somewhere, that the first value is overwritten?

### PR DESCRIPTION
**Pull request by dschwoerer**
resolves #1018 

I guess there are some cases where this will cause issues - e.g. my MMS tests in boutcore - but it might be worth it, as it could avoid issues of accidentally overwriting options ...

### CODE REVIEW

File: tests/integrated/test-subdir/data/BOUT.inp [hunk 1]
```diff
@@ -15,7 +15,7 @@ iyseps2_1	= -1
 iyseps2_2	= -1
 
 [All]
-bndry_all = neumann(0)
+#bndry_all = neumann(0)
```
Comment (bendudson) [inline tests/integrated/test-subdir/data/BOUT.inp]: Does this need to be commented out?

Comment (dschwoerer) [reply tests/integrated/test-subdir/data/BOUT.inp]: Removing is fine as well ...

**Comment by bendudson** [comment]
This is a good idea, and could catch some frustrating bugs early. I think we probably need a way to force a change of options, so the error can be suppressed or turned into a warning. I suggest either:

* Separate functions e.g. `reset` or `unsafeSet` or `forceSet` which are the same as `set` but without the throwing. They should still print a warning I think. Some code duplication though `set` could mostly just call the unsafe version
* A flag in Options which can be changed: `Options::allowMultipleDefinitions(true)`

File: src/sys/options.cxx [hunk 2]
```diff
@@ -60,8 +60,21 @@ void Options::set(const string &key, const string &val, const string &source) {
...
-  options[lowercase(key)] = opt;
+  string key_l = lowercase(key);
+  if (this->isSet(key_l)) {
+    const OptionValue & oldopt=options[key_l];
+    if (oldopt.value != opt.value) {
+      if (oldopt.source == opt.source) {
+        throw BoutException("Options: Setting a value from same source (%s) to new value %s - old value was %s.",source.c_str(),oldopt.value.c_str(),opt.value.c_str());
+      } else {
+        output_warn << "\tOption " << sectionName << ":" << key << " = " << oldopt.value
+                    << " (" << oldopt.source << ") overwritten with:" << endl
```
Comment (ZedThree) [inline src/sys/options.cxx]: Not a problem here, but just to note that `std::endl` also flushes to stdout immediately. `\n` is fine (and usually better) to use for a newline.

File: tests/unit/sys/test_options.cxx [hunk 3]
```diff
@@ -323,3 +323,14 @@ TEST_F(OptionsTest, MakeNestedSection) {
...
 }
+
+TEST_F(OptionsTest, SetSameOptionTwice) {
+  Options options;
+  options.set("key", "value", "code");
+  EXPECT_THROW(options.set("key", "new value", "code"),BoutException);
+  output_warn.disable();
+  options.set("key", "value", "code");
+  options.forceSet("key", "new value", "code");
+  options.set("key", "value", "code",true);
```
Comment (ZedThree) [inline tests/unit/sys/test_options.cxx]: Could you put `EXPECT_NO_THROW` on these two lines (333, 334)? Then if it does throw for some strange reason, googletest will catch it and carry on.

**Pull request merged**